File: examples/typescript/feedback-client.ts
// Minimal Feedback MCP client for web, Node.js, or React Native.
// Copy into your project and adjust the config.

export interface FeedbackClientConfig {
  /** Your instance, e.g. "https://feedback.your-domain.com" */
  endpoint: string;
  /** Project slug from config/projects/<slug> */
  project: string;
  /** The project's ingest key (safe to ship in clients; it can only submit) */
  ingestKey: string;
  /** "ios" | "android" | "web" | anything your project allows */
  platform?: string;
}

export interface SubmitFeedbackOptions {
  /** Form slug, e.g. "bug-report" */
  form: string;
  /** Must match the form's field schema */
  data: Record<string, unknown>;
  /** Anything useful for triage: app version, locale, device */
  metadata?: Record<string, unknown>;
  /** Optional end-user JWT if the project has auth.jwt configured */
  userToken?: string;
}

export class FeedbackError extends Error {
  constructor(
    public readonly status: number,
    message: string,
    public readonly details?: unknown
  ) {
    super(message);
    this.name = "FeedbackError";
  }
}

export async function submitFeedback(
  config: FeedbackClientConfig,
  options: SubmitFeedbackOptions
): Promise<{ id: string; createdAt: string }> {
  const response = await fetch(`${config.endpoint}/api/v1/feedback`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "X-Feedback-Key": config.ingestKey,
      ...(options.userToken ? { Authorization: `Bearer ${options.userToken}` } : {}),
    },
    body: JSON.stringify({
      project: config.project,
      form: options.form,
      platform: config.platform,
      data: options.data,
      metadata: options.metadata,
    }),
  });

  const body = await response.json().catch(() => ({}));
  if (!response.ok) {
    throw new FeedbackError(
      response.status,
      (body as { error?: string }).error ?? `Request failed with ${response.status}`,
      (body as { details?: unknown }).details
    );
  }
  return (body as { feedback: { id: string; createdAt: string } }).feedback;
}

// Usage:
//
// const feedback = await submitFeedback(
//   {
//     endpoint: "https://feedback.your-domain.com",
//     project: "my-app",
//     ingestKey: "pk_...",
//     platform: "web",
//   },
//   {
//     form: "bug-report",
//     data: { title: "Crash on launch", description: "...", severity: "high" },
//     metadata: { appVersion: "1.2.0", locale: navigator.language },
//   }
// );

File: examples/swift/FeedbackClient.swift
// Minimal Feedback MCP client for iOS, iPadOS, and macOS.
// Copy into your project and adjust the config.

import Foundation

struct FeedbackClient {
    /// Your instance, e.g. https://feedback.your-domain.com
    let endpoint: URL
    /// Project slug from config/projects/<slug>
    let project: String
    /// The project's ingest key (safe to ship in the binary; it can only submit)
    let ingestKey: String
    /// "ios", "macos", etc. Must be allowed by the project's platforms list.
    let platform: String

    enum FeedbackError: LocalizedError {
        case invalidResponse
        case server(status: Int, message: String?)

        var errorDescription: String? {
            switch self {
            case .invalidResponse:
                return "The feedback server returned an unexpected response."
            case .server(let status, let message):
                return message ?? "The feedback server returned status \(status)."
            }
        }
    }

    struct Submission: Decodable {
        let id: String
        let createdAt: String
    }

    /// Submits feedback. `data` must match the form's field schema.
    /// Pass `userToken` if the project has `auth.jwt` configured.
    @discardableResult
    func submit(
        form: String,
        data: [String: Any],
        metadata: [String: Any] = [:],
        userToken: String? = nil
    ) async throws -> Submission {
        var request = URLRequest(url: endpoint.appendingPathComponent("api/v1/feedback"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(ingestKey, forHTTPHeaderField: "X-Feedback-Key")
        if let userToken {
            request.setValue("Bearer \(userToken)", forHTTPHeaderField: "Authorization")
        }

        var defaultMetadata: [String: Any] = [
            "appVersion": Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") ?? "unknown",
            "build": Bundle.main.object(forInfoDictionaryKey: "CFBundleVersion") ?? "unknown",
            "osVersion": ProcessInfo.processInfo.operatingSystemVersionString,
            "locale": Locale.current.identifier,
        ]
        defaultMetadata.merge(metadata) { _, custom in custom }

        request.httpBody = try JSONSerialization.data(withJSONObject: [
            "project": project,
            "form": form,
            "platform": platform,
            "data": data,
            "metadata": defaultMetadata,
        ])

        let (body, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw FeedbackError.invalidResponse
        }
        guard (200..<300).contains(http.statusCode) else {
            let message = (try? JSONSerialization.jsonObject(with: body) as? [String: Any])?["error"] as? String
            throw FeedbackError.server(status: http.statusCode, message: message)
        }

        struct Envelope: Decodable { let feedback: Submission }
        return try JSONDecoder().decode(Envelope.self, from: body).feedback
    }
}

// Usage:
//
// let client = FeedbackClient(
//     endpoint: URL(string: "https://feedback.your-domain.com")!,
//     project: "my-app",
//     ingestKey: "pk_...",
//     platform: "ios"
// )
//
// try await client.submit(
//     form: "bug-report",
//     data: [
//         "title": "Crash on launch",
//         "description": "The app closes immediately after opening.",
//         "severity": "high",
//     ]
// )
